event-svc/src/event/validator/signed.rs

```diff
@@ -0,0 +1,267 @@
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+use ceramic_core::{SerializeExt, StreamId};
+use ceramic_event::unvalidated::{self, init, signed};
+use ceramic_validation::{cacao_verifier::Verifier, event_verifier::Verifier as _, VerifyJwsOpts};
+use cid::Cid;
+use ipld_core::ipld::Ipld;
+use tracing::warn;
+
+use crate::{
+    event::service::ValidationError,
+    store::{CeramicOneEvent, SqlitePool},
+    Result,
+};
+
+use super::{
+    event::{ValidatedEvent, ValidatedEvents},
+    grouped::{SignedEvents, Unsigned},
+    UnvalidatedEvent,
+};
+
+#[derive(Debug)]
+/// A signed validator is expected to be constructed to validate a batch of events and then be dropped.
+/// It will cache init events for the batch to use at different stages of validation and then drop everything
+/// when it finishes.
+pub struct SignedValidator {
+    /// The init events needed to validate this batch of events.
+    init_map: HashMap<Cid, Arc<unvalidated::Event<Ipld>>>,
+    /// The set of invalid init CIDs that can be used to reject any events that depend on them without even looking at them.
+    invalid_init: HashSet<Cid>,
+    /// The result of this validation batch that is updated as we go.
+    validated: ValidatedEvents,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum ValidationStatus {
+    ValidInit,
+    ValidData,
+    InvalidInit(String),
+    InvalidData(String),
+    Pending,
+}
+
+impl SignedValidator {
+    fn new(init_cnt: usize, data_cnt: usize) -> Self {
+        Self {
+            init_map: HashMap::with_capacity(init_cnt),
+            invalid_init: HashSet::new(),
+            validated: ValidatedEvents::new_with_expected_valid(init_cnt + data_cnt),
+        }
+    }
+
+    /// blindly trusts that this is indeed an init event
+    /// will error later if not
+    fn track_init(&mut self, cid: Cid, event: Arc<unvalidated::Event<Ipld>>) {
+        self.init_map.insert(cid, event);
+    }
+
+    fn add_invalid(&mut self, item: ValidationError) {
+        self.validated.invalid.push(item);
+    }
+
+    fn add_pending(&mut self, item: UnvalidatedEvent) {
+        self.validated.unvalidated.push(item);
+    }
+
+    fn add_valid(&mut self, item: ValidatedEvent) {
+        self.validated.valid.push(item);
+    }
+
+    fn add_valid_iter<I>(&mut self, item: I)
+    where
+        I: Iterator<Item = ValidatedEvent>,
+    {
+        self.validated.valid.extend(item);
+    }
+
+    pub async fn validate_events(
+        pool: &SqlitePool,
+        opts: &VerifyJwsOpts,
+        events: SignedEvents,
+        unsigned: Vec<Unsigned>,
```

Comment:
nit, its confusing that we are passing in events based on their signed status. I get that its helpful for the algorithm to know which is which. Maybe the `grouped` module can have a `Group` type that contains three sets of events, data, init, and unsigned init? That way this method has only a single parameter for the input events. This should help clarify the relationship between the events.

Reply:
Ah I found GroupedEvents in this PR. Why not use this type for the argument to this method?

Reply:
Yeah, I'll see what I can do. I originally just passed `SignedEvents` in, but Spencer (reasonably) found it confusing that SignedEvents contained unsigned init events, so I split it into three things (time, signed, unsigned) and passed two in here. I could probably take a reference to grouped, and maybe it's better since we _may_ want the time events to add timestamps someday. Currently those are being validated in parallel and I didn't want to fight over ownership but putting things in an Arc if needed is simple. Alternatively, I might be able to come up with a better name than `SignedEvents` and just pass that in here. 

The initial types/layout (like grouped) was split into #526, which this is targeting since this PR was getting big.

Reply:
I wanted to keep ownership of the events to make it easy for now, so I [renamed the types](https://github.com/ceramicnetwork/rust-ceramic/pull/503/commits/89ae7c556a3655c068117c04e571e548013c425c) on `GroupedEvents` `SignedValidationBatch` and `TimeValidationBatch` and pass those in.. lmk what you think.